### ISSUE TITLE
Re-implement JUnit 5 engine

### ISSUE THREAD

**Issue by marcphilipp** [junit-team/junit-lambda#39]
- [ ] Resolution (in #29)
- [x] ExtensionRegistry honors ordering
- [x] Implement getParent() of ExtensionContext
- [ ] Encapsulate attributes for ExtensionContext, do not expose Map
- [x] Invocation of ExtensionPoints
  - [x] BeforeEach
  - [x] AfterEach
  - [x] BeforeAll
  - [x] AfterAll
  - [x] PostProcessTestInstance
  - [X] MethodParameterResolver
  - [x] Conditional test execution
- [ ] Clean up integrated tests (subclasses of AbstractJUnit5TestEngineTestCase)

**Comment by mmerdes** [comment]
johannes and i will try to come up with a sketch of a new engine tomorrow.
should we use this issue or a different (yet to be created) one?

**Comment by marcphilipp** [comment]
I think referencing this issue is fine.

**Comment by mmerdes** [comment]
thanks for the clarification :)

**Comment by marcphilipp** [comment]
I'm working on cleaning up HierarchicalTestEngine.

**Comment by jlink** [comment]
I'm working on Before and After extension points.

**Comment by bechte** [comment]
Hi guys,

I have almost finished the implementation of the resolvers. Still there are some (minor) issues, that I would like to talk about:

1) Given a user has hand in an unique id to a method, the test cases require that both (the method descriptor and the class descriptor) are returned. According to JUnit4, this is a change. Do we really want to change this behavior? I somehow feel, that it is reasonable to run nodes that have been given by an unique id are run on the top level. Still not sure about this one…

From an execution’s point of view, our current implementation requires the test methods be be placed below a test class node, otherwise the beforeAll|Each/afterAll|Each will not get executed. Do we want this strong relationship? How about the package level? Shall we also introduce nodes for each package? This would be a consequence, at least with java 9… What do you think?


2) Handling of duplication: How do we want to handle duplication, i.e. if a test is specified more than once according to the hierarchy. Where do we want to place the test in the tree? Example:

TestClass {
  void test1();
  void test2();
}

Given a test plan specification with a class element „TestClass“ and an unique id element „junit5:TestCass.test1()“, how should the resulting tree look like? The most reasonable result I can think of is:

- junit5
-- TestClass
--- test1()
--- test2()


Based on these decisions, I will finally implement the handling of the duplication and will resolve the parent nodes if required for the method level. I would still like to integrate my branch as soon as possible, as it will conflict with mostly any of your changes by now, already. I pushed the branch "feature-29-extensible-discovery“ so you can review it. I would like to merge it very soon and continue to work on the open points on master, if that is fine with you…

Open points on my issue #29:
- Handling of duplication
- Handling of the tree building
- Optional: Adding package nodes for each package along the hierarchy
- Reactivate some ignored test cases
- Polish / Remove old/unnecessary classes



All the best,

Stefan

**Comment by jlink** [comment]
I'd like to have a thorough look at the branch since resolution is a critical feature. I think I'll be able to review it 2morrow night. 

As for the open questions:
- Always having all nodes down to the leaves seems reasonable to me since that simplifies execution of all beforeAll and afters in the hierarchy. An IDE can still decide to display the leaf only. 
- Unless we have some functionality on package level, we don't need them. As soon as we have they should always be present. The alternative would be to recreate all parents before execution which seems wasteful. 
- Eliminating duplication in the tree also looks like a good choice. 

Johannes

**Comment by bechte** [comment]
Hi Johannes

that’s fine, then I continue to work on the branch. I would like to finish up this task by Tuesday (night). So it would be great if some of you can have a look into it.

I will add the additional nodes along the hierarchy to simplify execution and eliminate the duplicated nodes as suggested.

Concerning the packages… We might want to rethink about this one (and discuss it on Tuesday) . If we add levels for the packages later, the IDE presentation „might" break, as they might implement some fancy logic to drill down and represent only single nodes (like in the „unique id for a method“ case, not showing the class container). If they do not expect the packages in the tree, their logic will somehow break. Therefore, we might want to explicitly add those nodes and keep the representation totally up to the tool vendors. This will guarantee that we don’t have to change anything on the test identifier tree.

Stefan

**Comment by marcphilipp** [comment]
I'm with Johannes on his answers to the open questions. I wouldn't add descriptors for packages at this point either. The IDEs need to consider nested tests anyway. I'm not that worried they'd break if we added an intermediate layer of descriptors later on.

**Comment by marcphilipp** [comment]
@bechte What happened to `NestedClassTestDescriptor`?

**Comment by bechte** [comment]
I removed it since the before and after methods are already handled in the ClassTestDescriptor (they were even executed twice) and I added a ReflectionUtils method to create a valid instance of a deep nested class (taken from my runner).

therefore, we have first level support for all kind of nested classes and need not to handle this case in a separate testdescriptor.

if you think it is worth separating these cases we could reintroduce the descriptor. 

best Stefan

**Comment by marcphilipp** [comment]
Are you sure they were executed twice? I thought we had a test that indicated they were only run once per test method.

Well, we introduced TestInstanceProvider to be able to later change when and how often test instances were created. Now, this is harder to change, I think.

**Comment by bechte** [comment]
the main difference is that the current implementation within the branch does not require a method test descriptor to be placed within a class test descriptor (see questions). 

Based on your answers I will change this behavior and with this change the need of an nested class test descriptor will return. :-)

I reimplemented everything step by step, test by test and maybe the remaining tests will point in this direction. but I wanted to clarify the questions before I continue to work. :-)

Stefan

**Comment by marcphilipp** [comment]
Sure, thanks for getting in touch! 👍

**Comment by jlink** [comment]
I've run into a tricky problem with TestExtensions and nested classes (which might be what Stefan meant when he wrote that before / after methods are executed twice, but maybe not):
Have a look at `MethodLevelCallbackTests`. If you remove the comments from the nested class, disable the first test and enable the second, the test failure message will show the problem: Test before/after extensions are invoked in both the context of the base class and the nested class which means that, for a single nested test method, foo and bar extensions are invoked twice :(
I saw no easy way to correct this behaviour, maybe someone else does?

**Comment by jlink** [comment]
With the help of @marcphilipp I could resolve the problem mentioned in the last comment. With pair programming many things would go a lot smoother :-/

**Comment by bechte** [comment]
I think we stumbled upon the same behavior from different sides.

I will also tackle it tonight and will have a look at your solution.

**Comment by jlink** [comment]
Have a Look at the lifecycle test cases. I THINK they specify the intended invocation order of before/after each.

**Comment by jlink** [comment]
I just had a look at branch feature29--extensible-discovery. Here's my comments in unspecified order:

1. I don't think the service loader approach doesn't pull its weight, since registering an external resolver does not make sense without having a corresponding Descriptor that fits exactly into the hierarchy of existing descriptors.

2. What is the intended usage of JUnit5TestEngine.setTestResolverRegistry() ?

3. I don't like the use of regular expressions in `ClassResolver` and `MethodResolver`. IMO they defy the purpose of composability of resolvers.

4. I liked `NestedDescriptor` as a concept to make the differences to `ClassDescriptor` explicit. The failing `MethodLevelCallbackTests` show that it's not easy to have the expected behaviour without a NestedDescriptor. I might be wrong, though.

5. In MethodLevelCallbackTests the engine creates one additional test descriptor for MethodLevelCallbackTests itself which is definitely not a part of the test hierarchy.

1, 3 and 5 are critical issues for me that should be resolved. As for the missing `NestedClassDescriptor`: I'm happy without it iff we can create the desired behaviour (as specicifed by callSequence in MethodLevelCallbackTests`).

Suggestion to Stefan: If you can resolve my (and other team members') issues, make a pull request that we can then merge tomorrow.

**Comment by bechte** [comment]
Hi Johannes,

thanks for reviewing and your suggestions. I modified some points but let's go them one by one:

1. Actually, I really think it does. A TestResolver will always come in pair with a TestDescriptor. In a hard-coded approach there would be no way to introduce new TestDescriptors without changing existing code. I agree, that there is a tight coupling between the MethodResolver and the ClassResolver, but I think we can easily circumstance this by introducing some kind of "Path" representing the unique id as proposed by @marcphilipp. I haven't work on that, because I wanted to have something we can continue to work on quickly. Also, I think we should extract the concept of a NestedDescriptor. I totally agree that we should have this, but I would like to extract this as a first "extension" of the resolver mechanism. Again, I wanted to have something, that we can merge into master and continue to work on separate parts.

2. I removed the method. It was originally used in a test which was replaced by your SpecificationResolverTests. Therefore, I simply removed it, because it is no longer used.

3. I don't like the use of regular expressions in this case either. But for me this is an issue of the missing "Path" concept and has nothing to do with the concept of a resolver. We had the same concept of splitting the unique id within the JUnit5TestableFactory. As soon as we introduce a valid separation of the path elements, we will have a way to remove the regular expressions as well.

4. As mentioned above I also like the NestedClassTestDescriptor approach. The test was not active in my implementation as I worked on the master and your work on before/after methods has not been rebased. The test now fail after I rebased my branch on the master. Therefore, I need to continue to work on this, but I needed your changes before I could do so. I will fix this now and check it into the feature branch, so we can review this tomorrow.

5. We discussed this point in the skype telco today. I totally agree with you and so did the others, i.e. inner static classes must be treated like top-level classes. This is also what my "HierarchicalContextRunner" does. I just wanted to discuss this, before making the change. I will also change this today so you can have another look tomorrow.

So in summary: Issues 2, 4 and 5 as solved. Issue 3 needs some more investigation on the "path" concept, which we might want to do in a pairing in January (or anytime). Issue 1 is not an issue but an advantage to me. With the concept of a "path" it will also allow us to introduce package level nodes (with Java 9) without touching the current code base. This is an absolute pro, plus we allow extension writers to add additional nodes by providing a resolver. We can also reuse this mechanism for providing dynamic tests... Therefore, I definitely want to keep 1 as it is.

I created a pull request for this merge: #65

Best Stefan

**Comment by jlink** [comment]
Did some experimentation to build in conditional test execution with extension points. Since this will require changes to `HierarchicalTestEngine` I'd rather do that together with someone else.

**Comment by jlink** [comment]
Created new issues for remaining tasks:
- Specification resolution in #29 
- Clean up integrated engine tests in #81
- Encapsulate attributes for ExtensionContext in #82
